### ISSUE TITLE
InvalidCastException when serializing IDictionary<TKey,TValue>, IDictionary implementation

### ISSUE THREAD

**Issue by KimikoMuffin** [JamesNK/Newtonsoft.Json#501]
I have a project that uses a dictionary class which implements both `System.Collections.Generic.IDictionary<TKey,TValue>` and `System.Collections.IDictionary`. When I try to serialize this dictionary, it throws an InvalidCastException on line 915 of JsonSerializerInternalWriter.cs, which is, "`foreach (DictionaryEntry entry in values)`", since the contents of the dictionary are "supposed to be" KeyValuePair<TKey, TValue>.

I realize the optimal solution is "don't implement `System.Collections.IDictionary`", but I figured you should at least be aware of this.

For that matter, the standard `System.Collections.Generic.Dictionary<TKey, TValue>` itself doesn't cause this error, for some reason.

**Comment by JamesNK** [comment]
Dictionaries that don't implement IDictionary will be wrapped. Upload a repro of the type you're serializing and I'll look at it.

**Comment by KimikoMuffin** [comment]
https://spideroak.com/browse/share/KimikoMuffin/JsonTest This project contains everything relevant.

**Comment by KimikoMuffin** [comment]
... Eureka, probably.

On line 121 of WhatDictionary.cs in the project I just linked, it implements ``System.Collections.IDictionary.GetEnumerator()``, by just calling ``_dict.GetEnumerator()`` on a ``System.Collections.Generic.Dictionary<TKey,TValue>``. However, when *that* class implements ``System.Collections.IDictionary``, it uses a *completely different enumerator-type*, one which only implements ``System.Collections.IDictionaryEnumerator``. The problem was solved when I changed line 123 of WhatDictionary.cs to ``((IDictionary)_dict).GetEnumerator()``

So, bottom line is, the InvalidCastException was my fault. ;)
